docs/adr/0004-workflow-instance-indexing-and-trigger-acceptance.md

```diff
@@ -0,0 +1,61 @@
+
+In the scenario where a trigger targets a number of workflow instances based on their interests, the change in approach does not introduce any new problems; there is currently, and will continue to be, a challenge for the client to ensure that they take into account the type of consistency issues described above.
+
+In the scenario where a trigger is intended to target a specific workflow instance using its Id, it doesn't actually matter whether the index is stale or not as long as it's been created. This is because a workflow instance Id is immutable and will always be part of the interests collection so will always be in the index.
+
+The only scenario that is problematic is (as before) during workflow instance creation. We can break this down into two:
+- a trigger is raised by an action executed as part of workflow initialization
+- a client needs to wait for a workflow instance creation request to complete and then immediately send a trigger.
+
+In the first case, we need to ensure the index entry for the new instance has been created prior to raising the new triggers. The only way of making this the case is to make updating the index and raising the triggers part of the same process. To achieve this, we can implement a way for an action to return a trigger for the workflow engine to create at the appropriate point.
+
+The second case is more problematic. In this scenario it is possible that the trigger will be received before the index contains the `Interests` of the new workflow instance.
+
+This scenario is not considered to be high priority. A scenario where a client needs to wait for workflow instance creation to complete and then immediately send a trigger is possible but unlikely, and would suggest an issue with workflow design. In addition, if it really is necessary for a client to be able to do this, it can be built into the workflow design by introducing an "initializing" workflow state to the start of a workflow. The client can then wait for the instance to leave the "initializing" state, at which point it can be certain that the index exists, and can send any further triggers.
+
+## Decision
+
+At the time of writing (and as part of the changes to move to the event sourcing approach) workflow actions are able to return updates to be made to a workflow instance `Context`. We will extend this to allow the action to return 0 or more triggers that the engine will execute once the workflow instance index of interests is up to date.
+
+We will extend the events raised as part of a transition to include a final `WorkflowInstanceTransitionComplete` event which will include any triggers to be raised once the workflow instance updates are complete and the instance index of interests is updated. We will have an event stream processor which uses this event to update the workflow instance index and dispatches these triggers.
```

Comment:
I think that you can argue that this is a better solution all round. The engine basically becomes responsible for raising the right Workflow Engine events in the right order, and things observe those events and carry out the actions concerned.

This re-raises the question: should this be how we do the rest of it, too? The actions events should be raised and processed to carry out the actions, in the same way?

Reply:
I would say not - because an event indicates that "something has happened", and the actions are how we "make the thing happen". So we can't say that something has happened until we've executed the actions.

However, I think some (maybe most of) the things we currently implement as actions should be re-implemented by handling events coming out of the workflow engine. E.g. anything that builds a projection should be doing it by processing events rather than in actions. This is desirable anyway, because it allows the projections to be rebuilt by processing the event stream.

We did discuss a while back the idea of getting rid of the external actions completely and saying that actions can only work off a combination of data in the trigger and data attached to the worfklow context. This would potentially make the the workflow engine much faster and more reliable, at the same time encouraging "good behaviour" when designing workflows.